hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/HoodieSparkMergeHandleV2.java

```diff
@@ -0,0 +1,410 @@
+  }
+
+  protected void makeOldAndNewFilePaths(String partitionPath, String oldFileName, String newFileName) {
+    oldFilePath = makeNewFilePath(partitionPath, oldFileName);
+    newFilePath = makeNewFilePath(partitionPath, newFileName);
+  }
+
+  // TODO(yihua): is this still needed
+  /*
+  if (needsUpdateLocation()) {
+        record.unseal();
+        record.setNewLocation(new HoodieRecordLocation(instantTime, fileId));
+        record.seal();
+      }
+   */
+
+  public void write() {
+    boolean usePosition = config.getBooleanOrDefault(MERGE_USE_RECORD_POSITIONS);
+    Schema readerSchema;
+    Option<InternalSchema> internalSchemaOption = Option.empty();
+    if (!StringUtils.isNullOrEmpty(config.getInternalSchema())) {
+      readerSchema = HoodieAvroUtils.addMetadataFields(
+          new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
+      internalSchemaOption = SerDeHelper.fromJson(config.getInternalSchema());
+    } else {
+      readerSchema = HoodieAvroUtils.addMetadataFields(
+          new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
+    }
+    // TODO(yihua): reader schema is good enough for writer?
+    HoodieFileGroupReader<T> fileGroupReader = new HoodieFileGroupReader<>(
+        readerContext,
+        storage,
+        config.getBasePath(),
+        instantTime,
+        fileSlice,
+        readerSchema,
+        readerSchema,
+        internalSchemaOption,
+        hoodieTable.getMetaClient(),
+        hoodieTable.getMetaClient().getTableConfig().getProps(),
+        0,
+        Long.MAX_VALUE,
+        usePosition);
+    try {
+      fileGroupReader.initRecordIterators();
+      HoodieFileGroupReaderIterator<InternalRow> recordIterator
+          = (HoodieFileGroupReaderIterator<InternalRow>) fileGroupReader.getClosableIterator();
+      StructType sparkSchema = AvroConversionUtils.convertAvroSchemaToStructType(readerSchema);
+      while (recordIterator.hasNext()) {
+        InternalRow row = recordIterator.next();
+        HoodieKey recordKey = new HoodieKey(
+            row.getString(HoodieRecord.COMMIT_SEQNO_METADATA_FIELD_ORD),
+            row.getString(HoodieRecord.PARTITION_PATH_META_FIELD_ORD));
+        HoodieSparkRecord record = new HoodieSparkRecord(recordKey, row, sparkSchema, false);
+        Option recordMetadata = record.getMetadata();
+        if (!partitionPath.equals(record.getPartitionPath())) {
+          HoodieUpsertException failureEx = new HoodieUpsertException("mismatched partition path, record partition: "
+              + record.getPartitionPath() + " but trying to insert into partition: " + partitionPath);
+          writeStatus.markFailure(record, failureEx, recordMetadata);
+          continue;
+        }
+        try {
+          writeToFile(recordKey, record, readerSchema, config.getPayloadConfig().getProps(), preserveMetadata);
```

Comment:
does this handle all sort of delete/payload handling - `_hoodie_is_deleted`.. and emptypayload..

Reply:
should be handled by the fg reader itself, I think.

Reply:
This is fixed now.